node-gyp build error "missing any Windows SDK"

### Current behavior

Executing `yarn` in a local copy of the [cypress repo](https://github.com/cypress-io/cypress) on Windows 10 or 11 causes warnings on installing optional dependencies:

- [registry-js](https://www.npmjs.com/package/registry-js)
- [cpu-features](https://www.npmjs.com/package/cpu-features)

[node-gyp](https://www.npmjs.com/package/node-gyp) outputs the error message "missing any Windows SDK" from [Visual Studio 2022](https://visualstudio.microsoft.com/vs/) so the packages fail to install.

Yarn is using [node-gyp@8.4.1](https://github.com/nodejs/node-gyp/releases/tag/v8.4.1) to install the above packages.

Support for Windows 11 SDK was however not added until [node-gyp@9.1.0](https://github.com/nodejs/node-gyp/releases/tag/v9.1.0).

The instructions [node-gyp > Installation > On Windows](https://github.com/nodejs/node-gyp/blob/main/README.md#on-windows) do not work for the current version of Cypress:

> Install the current [version of Python](https://devguide.python.org/versions/) from the [Microsoft Store](https://apps.microsoft.com/store/search?publisher=Python+Software+Foundation).
>
> Install tools and configuration manually:
>   * Install Visual C++ Build Environment: [Visual Studio Build Tools](https://visualstudio.microsoft.com/thank-you-downloading-visual-studio/?sku=BuildTools)
>   (using "Visual C++ build tools" if using a version older than VS2019, otherwise use "Desktop development with C++" workload) or  [Visual Studio Community](https://visualstudio.microsoft.com/thank-you-downloading-visual-studio/?sku=Community) (using the "Desktop development with C++" workload)

### Desired behavior

[CONTRIBUTING > Requirements](https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#requirements) should correctly list working prerequisites. The current text says:

> - [`python`](https://www.python.org/downloads/) (since we use `node-gyp`. See their [repo](https://github.com/nodejs/node-gyp) for Python version requirements.)

1. Issue https://github.com/cypress-io/cypress/issues/28695 should be resolved by requiring that Python version <= `3.11` be installed.
2. The instructions should say that the Windows 10 SDK needs to be installed (since the Windows 11 SDK cannot be detected).

### Test code to reproduce

- Install [Python 3.11](https://devguide.python.org/versions/) from the [Microsoft Store](https://apps.microsoft.com/detail/9NRWMJP3717K?hl=en-us&gl=US).

- Install [Visual Studio Community](https://visualstudio.microsoft.com/thank-you-downloading-visual-studio/?sku=Community) version 2022, using the workload "Desktop development with C++" with its default selections (which includes Windows 11 SDK).

```shell
git clone https://github.com/cypress-io/cypress
```

Use Node.js `18.15.0`

If testing on existing clone then

```shell
git switch develop
git clean -x -f -d
```

then execute

```shell
yarn
```


### Cypress Version

`13.6.2` or commit https://github.com/cypress-io/cypress/commit/09ce5c1f0c4920d820f10b96666cb3a7f259000a

### Node version

`v18.15.0`

### Operating System

Microsoft Windows 10 and Windows 11

### Debug Logs

```shell
[5/6] Building fresh packages...
warning Error running install script for optional dependency: "C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\cpu-features: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\cpu-features
Output:
gyp ERR! find VS
gyp ERR! find VS msvs_version not set from command line or npm config
gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
gyp ERR! find VS checking VS2022 (17.8.34408.163) found at:
gyp ERR! find VS \"C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\"
gyp ERR! find VS - found \"Visual Studio C++ core features\"
gyp ERR! find VS - found VC++ toolset: v143
gyp ERR! find VS - missing any Windows SDK
gyp ERR! find VS could not find a version of Visual Studio 2017 or newer to use
gyp ERR! find VS looking for Visual Studio 2015
gyp ERR! find VS - not found
gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
gyp ERR! find VS
gyp ERR! find VS **************************************************************
gyp ERR! find VS You need to install the latest version of Visual Studio
gyp ERR! find VS including the \"Desktop development with C++\" workload.
gyp ERR! find VS For more information consult the documentation at:
gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
gyp ERR! find VS **************************************************************
gyp ERR! find VS
gyp ERR! configure error
gyp ERR! stack Error: Could not find any Visual Studio installation to use
gyp ERR! stack     at VisualStudioFinder.fail (C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:122:47)
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:75:16
gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:363:14)
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:71:14
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:384:16
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\util.js:54:7
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\util.js:33:16
gyp ERR! stack     at ChildProcess.exithandler (node:child_process:427:5)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at maybeClose (node:internal/child_process:1091:16)
gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
gyp ERR! System Windows_NT 10.0.19045
gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\mikem\\\\github\\\\cypress-io\\\\cypress\\\\node_modules\\\\@npmcli\\\\run-script\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"rebuild\"
gyp ERR! cwd C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\cpu-features
gyp ERR! node -v v18.15.0
gyp ERR! node-gyp -v v8.4.1
gyp ERR! not ok"
info This module is OPTIONAL, you can safely ignore this error
warning Error running install script for optional dependency: "C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@cypress\\get-windows-proxy\\node_modules\\registry-js: Command failed.
Exit code: 1
Command: prebuild-install || node-gyp rebuild
Arguments:
Directory: C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@cypress\\get-windows-proxy\\node_modules\\registry-js
Output:
prebuild-install WARN install No prebuilt binaries found (target=18.15.0 runtime=node arch=x64 libc= platform=win32)
gyp ERR! find VS
gyp ERR! find VS msvs_version not set from command line or npm config
gyp ERR! find VS VCINSTALLDIR not set, not running in VS Command Prompt
gyp ERR! find VS checking VS2022 (17.8.34408.163) found at:
gyp ERR! find VS \"C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\"
gyp ERR! find VS - found \"Visual Studio C++ core features\"
gyp ERR! find VS - found VC++ toolset: v143
gyp ERR! find VS - missing any Windows SDK
gyp ERR! find VS could not find a version of Visual Studio 2017 or newer to use
gyp ERR! find VS looking for Visual Studio 2015
gyp ERR! find VS - not found
gyp ERR! find VS not looking for VS2013 as it is only supported up to Node.js 8
gyp ERR! find VS
gyp ERR! find VS **************************************************************
gyp ERR! find VS You need to install the latest version of Visual Studio
gyp ERR! find VS including the \"Desktop development with C++\" workload.
gyp ERR! find VS For more information consult the documentation at:
gyp ERR! find VS https://github.com/nodejs/node-gyp#on-windows
gyp ERR! find VS **************************************************************
gyp ERR! find VS
gyp ERR! configure error
gyp ERR! stack Error: Could not find any Visual Studio installation to use
gyp ERR! stack     at VisualStudioFinder.fail (C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:122:47)
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:75:16
gyp ERR! stack     at VisualStudioFinder.findVisualStudio2013 (C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:363:14)
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:71:14
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\find-visualstudio.js:384:16
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\util.js:54:7
gyp ERR! stack     at C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@npmcli\\run-script\\node_modules\\node-gyp\\lib\\util.js:33:16
gyp ERR! stack     at ChildProcess.exithandler (node:child_process:427:5)
gyp ERR! stack     at ChildProcess.emit (node:events:513:28)
gyp ERR! stack     at maybeClose (node:internal/child_process:1091:16)
gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
gyp ERR! System Windows_NT 10.0.19045
gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\mikem\\\\github\\\\cypress-io\\\\cypress\\\\node_modules\\\\@npmcli\\\\run-script\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"rebuild\"
gyp ERR! cwd C:\\Users\\mikem\\github\\cypress-io\\cypress\\node_modules\\@cypress\\get-windows-proxy\\node_modules\\registry-js
gyp ERR! node -v v18.15.0
gyp ERR! node-gyp -v v8.4.1
gyp ERR! not ok"
info This module is OPTIONAL, you can safely ignore this error
```

### Other

Installing a Windows 10 SDK prevents the error message "missing any Windows SDK", however the installation still fails.

## Windows 10 

installation of `cpu-features` fails with:
>  'cmake' is not recognized as an internal or external command,
 > operable program or batch file.

installation of `registry-js` fails with:
> error C1083: Cannot open include file: 'v8-locker.h': No such file or directory 

## Windows 11

installation of `cpu-features` fails with:
>  'cmake' is not recognized as an internal or external command,
 > operable program or batch file.

no errors from installation of `registry-js`

---

At this time I have found no solution to installing (rebuilding) Cypress from source on Windows 10 and 11. Yarn Classic also times out if a high value for `yarn config set network-timeout 600000` is not set.

I'm not very hopeful to find any satisfactory workaround. 🙁

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -207,6 +207,7 @@ You must have the following installed on your system to contribute locally:
 - [`yarn`](https://yarnpkg.com/en/docs/install)
 - [`python`](https://www.python.org/downloads/) (since we use `node-gyp`. See their [repo](https://github.com/nodejs/node-gyp) for Python version requirements.)
   - Note for Debian-based systems: `python` is pre-installed.<br>`sudo apt install g++ make cmake` meets the additional requirements to run `node-gyp` in the context of building Cypress from source.
+  - Note for Windows systems: when installing the Visual Studio C++ environment recommended by [node-gyp](https://github.com/nodejs/node-gyp), install also a Windows 10 SDK. The currently used version of `node-gyp` may otherwise fail to recognise the Visual Studio installation.
 
 ### Getting Started
 

